### PR TITLE
Add vouching

### LINKED ISSUE #1093: Allow Good Status Uploaders to approve other uploaders.
Arxiv or lobste.rs style -- as suggested by chris martin on twitter.

Draft workflow: once someone registers an account they get a magic link that they can send to two people who are themselves approved uploaders for A) more than a month and B) have at least two uploads each. Those two can click the magic link to approve the new uploader.

This alleviates the burden from trustees. @ysangkok @andreasabel  what do you think?

### PR DESCRIPTION
**Pull request by ysangkok**
Fixes #1093.

Vouching allows new users to get added to the uploaders group by way of other people "vouching" for then. This should alleviate privileged Hackage users, since they were previously the only people that could add people to the uploaders group.

Tagging @david-christiansen since we discussed this.

### CODE REVIEW

File: src/Distribution/Server/Features/Vouch.hs [hunk 1]
```diff
@@ -0,0 +1,218 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE DerivingStrategies #-}
+module Distribution.Server.Features.Vouch where
```
Comment (jappeace) [inline src/Distribution/Server/Features/Vouch.hs:6]: shouldn't u do explict exports?

File: src/Distribution/Server/Features/Vouch.hs [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+renderToLBS lookupUserInfo vouches = do
+  rendered <- traverse renderVouchers vouches
+  pure $
+    templateUnescaped "vouches" $
+      if null rendered
+         then LBS.pack "Nobody has vouched yet."
+         else LBS.intercalate mempty rendered
+  where
+  renderVouchers :: (UserId, UTCTime) -> ServerPartE LBS.ByteString
+  renderVouchers (uid, timestamp) = do
```
Comment (jappeace) [inline src/Distribution/Server/Features/Vouch.hs:138]: I'd put these local scoped functions as top level bindings.

File: src/Distribution/Server/Features/Vouch.hs [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+data Err
+  = NotAnUploader
+  | You'reTooNew
+  | VoucheeAlreadyUploader
+  | AlreadySufficientlyVouched
+  | YouAlreadyVouched
+
+data Success = AddVouchComplete | AddVouchIncomplete
+
+judge :: Group.UserIdSet -> UTCTime -> UserId -> [(UserId, UTCTime)] -> [(UserId, UTCTime)] -> UserId -> Either Err (Either Err Success)
```
Comment (jappeace) [inline src/Distribution/Server/Features/Vouch.hs:109]: why does this need 2 levels of Either?

File: src/Distribution/Server/Features/Vouch.hs [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+          ]
+      handlePostVouch :: DynamicPath -> ServerPartE Response
+      handlePostVouch dpath = do
+        voucher <- guardAuthenticated
+        ugroup <- liftIO $ Group.queryUserGroup uploadersGroup
+        now <- liftIO getCurrentTime
+        vouchee <- lookupUserName =<< userNameInPath dpath
+        vouchersForVoucher <- queryState vouchState $ GetVouchesFor voucher
+        existingVouchers <- queryState vouchState $ GetVouchesFor vouchee
+        case join $ judge ugroup now vouchee vouchersForVoucher existingVouchers voucher of
```
Comment (jappeace) [inline src/Distribution/Server/Features/Vouch.hs:173]: Why pawn off all checks to a judge function to immediatly unwrap them again?

File: src/Distribution/Server/Features/Vouch.hs [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+  getFeatureInterface = vouchFeatureInterface
+
+requiredCountOfVouches :: Int
+requiredCountOfVouches = 3
+
+isWithinLastMonth :: UTCTime -> (UserId, UTCTime) -> Bool
+isWithinLastMonth now (_, vouchTime) =
+  addUTCTime (30 * nominalDay) vouchTime < now
+
+data Err
```
Comment (jappeace) [inline src/Distribution/Server/Features/Vouch.hs:100]: I'd write this out as a full word, eg VouchError, but I think the entire data structure and judge function is superflous.

File: src/Distribution/Server/Features/Vouch.hs [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+data VouchFeature =
+  VouchFeature
+    { vouchFeatureInterface :: HackageFeature
+    }
+
+instance IsHackageFeature VouchFeature where
+  getFeatureInterface = vouchFeatureInterface
+
+requiredCountOfVouches :: Int
+requiredCountOfVouches = 3
```
Comment (jappeace) [inline src/Distribution/Server/Features/Vouch.hs:94]: seems like a lot of vouches  to require, I don't think I would've gotten this number when working at daisee, only one of my colleagues had access to hackage back then.

File: src/Distribution/Server/Features/Vouch.hs [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+  renderVouchers (uid, timestamp) = do
+    info <- lookupUserInfo uid
+    let UserName name = userName info
+        -- We don't need to show millisecond precision
+        -- So we truncate it off here
+        truncated = truncate $ utctDayTime timestamp
+        newUTCTime = timestamp {utctDayTime = secondsToDiffTime truncated}
+    pure . toUTF8LBS . prettyHtmlFragment . li . stringToHtml $ name <> " vouched on " <> formatShow iso8601Format newUTCTime
+
+initVouchFeature :: ServerEnv -> IO (UserFeature -> UploadFeature -> IO VouchFeature)
```
Comment (jappeace) [inline src/Distribution/Server/Features/Vouch.hs:147]: Is this callback setup required? can't you immediatly chug in UserFeatue and UploadFeature?

eg 
```
initVouchFeature :: ServerEnv -> UserFeature -> UploadFeature -> IO VouchFeature
```

**Review by jappeace — COMMENTED** [review]
doesn't hackage do tests either? LIke I'd expect at least all happy paths to have some integration test

File: src/Distribution/Server/Features/Vouch.hs (hunk 7 above)
Comment (ysangkok) [reply src/Distribution/Server/Features/Vouch.hs:147]: This is the pattern used across the codebase, look in `Server/Features.hs` (the use site) so see how it's used all over. I don't object to your observation but I think we should change it consistently in a different PR if it needs to change.

File: src/Distribution/Server/Features/Vouch.hs (hunk 6 above)
Comment (ysangkok) [reply src/Distribution/Server/Features/Vouch.hs:94]: Reducing to 2.

File: src/Distribution/Server/Features/Vouch.hs (hunk 4 above)
Comment (ysangkok) [reply src/Distribution/Server/Features/Vouch.hs]: This is primarily such that the logic can be tested. I added tests in the latest comment.

File: src/Distribution/Server/Features/Vouch.hs (hunk 7 above)
Comment (jappeace) [reply src/Distribution/Server/Features/Vouch.hs:158]: it's fine, it's a minor style nitpick :)

**Review by jappeace — APPROVED** [review]
You've my approval although I know little about hackage

**Comment by gbaz** [comment]
I'll give this a real look soon and edit the text a bit as well, but just as a note i think the right name for the text is endorsement, following arxiv https://info.arxiv.org/help/endorsement.html

Of course in the code itself we can keep vouch to not cause a bunch of renames :-P

**Comment by ysangkok** [comment]
@gbaz I added commits to add an e-mail notification when the endorsements have been received and to rename to user-facing pages to refer to 'endorsements'.

**Pull request merged**